src/celeritas/global/ActionLauncher.device.hh

```diff
@@ -70,17 +66,24 @@ class ActionLauncher
                   "Launched action must be a trivially copyable function "
                   "object");
 
+  private:
+    using kernel_func_ptr_t = void (*)(Range<ThreadId> const, F);
+
+    // TODO: better way to conditionally constexpr init?
+    static constexpr kernel_func_ptr_t kernel_func_ptr
+        = detail::select_kernel<F>();
```

Comment:
@sethrj I was thinking, maybe, instead of having select_kernel return a pointer to the correct kernel, we can instead instantiate a single kernel. I'm not sure there would be any performance impact though since the compiler can figure out which kernel we want to call at compile time it probably optimizes out the extra indirection of de-referencing the pointer.

Reply:
You're probably right. Might be good to check though?

Reply:
very boiled down example but the constexpr function pointer results in a direct call to the function without needing to load the address in a register: https://godbolt.org/z/Thf5zsMzb

Reply:
Nice. I think I may have misunderstood you: I was concerned with NVCC resulting in different occupancy from putting two function calls  (one `if constexpr`'d out) in one kernel. But that's clearly not what you're doing. The new code looks great either way. 🚢

Reply:
Oh yeah not exactly what I had in mind. It took a few iterations but we ended up converging on a pretty concise implementation.